### PR TITLE
python3Packages.commitizen: Add library along with binary

### PR DESCRIPTION
**Pull request by shivaraj-bh**
custom commit rules require the library, see:
https://commitizen-tools.github.io/commitizen/customization/#2-customize-through-customizing-a-class

cc: @lovesegfault @anthonyroussel 
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by anthonyroussel** [comment]
Result of `nixpkgs-review pr 329638` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>commitizen (python312Packages.commitizen)</li>
    <li>commitizen.dist (python312Packages.commitizen.dist)</li>
  </ul>
</details>

**Comment by anthonyroussel** [comment]
@ofborg build commitizen python3Packages.commitizen

**Comment by anthonyroussel** [comment]
Your commit message didn't trigger automatically the ofborg CI.
I launched it manually using the "ofborg build" command.

Your commit message should be:

> python312Packages.commitizen: Add library along with binary

so that Ofborg CI automatically detects the built package.

See https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building

**Review by anthonyroussel — APPROVED** [review]

File: pkgs/development/python-modules/commitizen/default.nix [hunk 1]
```diff
@@ -9,7 +9,7 @@
 , testers
 }:
 
-python3.pkgs.buildPythonApplication rec {
+python3.pkgs.buildPythonPackage rec {
```
Comment (anthonyroussel) [inline pkgs/development/python-modules/commitizen/default.nix:12]: Should be

```suggestion
buildPythonPackage rec {
```

for Python library.
You need to import buildPythonPackage and all the Python dependencies.

Comment (anthonyroussel) [reply pkgs/development/python-modules/commitizen/default.nix:12]: Same for nativeBuildInputs, propagatedBuildInputs, nativeCheckInputs.

Example:

```
nativeBuildInputs = with python3.pkgs; [
  poetry-core
  installShellFiles
];
```

should become for Python library:

```nix
{
  argcomplete,
  charset-normalizer,
  colorama,
  decli,
  importlib-metadata,
  jinja2,
  packaging,
  pyyaml,
  questionary,
  termcolor,
  tomlkit,
}:

propagatedBuildInputs = with python3.pkgs; [
  argcomplete
  charset-normalizer
  colorama
  decli
  importlib-metadata
  jinja2
  packaging
  pyyaml
  questionary
  termcolor
  tomlkit
];
```

Comment (shivaraj-bh) [reply pkgs/development/python-modules/commitizen/default.nix]: Done in https://github.com/NixOS/nixpkgs/pull/329638/commits/3b1b6a1d40413af633271c63b18af26152623652

**Comment by shivaraj-bh** [comment]
> Your commit message didn't trigger automatically the ofborg CI. I launched it manually using the "ofborg build" command.
> 
> Your commit message should be:
> 
> > python312Packages.commitizen: Add library along with binary
> 
> so that Ofborg CI automatically detects the built package.
> 
> See https://github.com/NixOS/ofborg?tab=readme-ov-file#automatic-building

Changed it now

**Review by anthonyroussel — APPROVED** [review]
Result of `nixpkgs-review pr 329638` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>commitizen (python312Packages.commitizen)</li>
    <li>commitizen.dist (python312Packages.commitizen.dist)</li>
    <li>python311Packages.commitizen</li>
    <li>python311Packages.commitizen.dist</li>
  </ul>
</details>

Package now build successfully for python311

File: pkgs/development/python-modules/commitizen/default.nix
Comment (anthonyroussel) [inline pkgs/development/python-modules/commitizen/default.nix:1]: I forgot, we need to do the same thing for `argcomplete` in the `postInstall` and `pythonOlder`:

```diff
diff --git a/pkgs/development/python-modules/commitizen/default.nix b/pkgs/development/python-modules/commitizen/default.nix
index 10171ddf1d2e..25b10b31f452 100644
--- a/pkgs/development/python-modules/commitizen/default.nix
+++ b/pkgs/development/python-modules/commitizen/default.nix
@@ -3,7 +3,7 @@
 , fetchFromGitHub
 , buildPythonPackage
 , git
-, python3
+, pythonOlder
 , stdenv
 , installShellFiles
 , poetry-core
@@ -33,7 +33,7 @@ buildPythonPackage rec {
   version = "3.27.0";
   format = "pyproject";
 
-  disabled = python3.pythonOlder "3.8";
+  disabled = pythonOlder "3.8";
 
   src = fetchFromGitHub {
     owner = "commitizen-tools";
@@ -106,14 +106,14 @@ buildPythonPackage rec {
 
   postInstall =
     let
-      argcomplete = lib.getExe' python3.pkgs.argcomplete "register-python-argcomplete";
+      register-python-argcomplete = lib.getExe' argcomplete "register-python-argcomplete";
     in
     lib.optionalString (stdenv.buildPlatform.canExecute stdenv.hostPlatform)
       ''
         installShellCompletion --cmd cz \
-          --bash <(${argcomplete} --shell bash $out/bin/cz) \
-          --zsh <(${argcomplete} --shell zsh $out/bin/cz) \
-          --fish <(${argcomplete} --shell fish $out/bin/cz)
+          --bash <(${register-python-argcomplete} --shell bash $out/bin/cz) \
+          --zsh <(${register-python-argcomplete} --shell zsh $out/bin/cz) \
+          --fish <(${register-python-argcomplete} --shell fish $out/bin/cz)
       '';
 
   passthru = {

```

Comment (shivaraj-bh) [reply pkgs/development/python-modules/commitizen/default.nix:1]: Right, thanks for the patch. Used it in https://github.com/NixOS/nixpkgs/pull/329638/commits/ed3cc73c45957a7b7ca074fa60b4d213d3a274cb

**Review by anthonyroussel — APPROVED** [review]

**Pull request merged**